### PR TITLE
add events fetch per one request

### PR DESCRIPTION
**Pull request by mike-lvov**
Fixes the issue with thousands of events requests.

NOTE: We should also create an index

image

### CODE REVIEW

**Comment by 0xdevalias** [comment]
@yarikoptic @soichih FYI

File: src/types/venues.ts [hunk 1]
```diff
@@ -319,6 +319,7 @@ export interface VenueEvent {
   collective_price: number;
   host: string;
   room?: string;
+  venueId?: string;
```
Comment (0xdevalias) [inline src/types/venues.ts:322]: Is this going to be on every event always? We have a WithVenueId helper.

File: firestore.rules [hunk 2]
```diff
@@ -48,13 +52,13 @@ service cloud.firestore {
       allow read: if request.auth.uid != null;
       allow write: if true;
     }
+    match /{path=**}/events/{event} {
+      allow read: if true;
+      allow write: if request.auth.uid in venueData().owners;
+    }
```
Comment (mike-lvov) [inline firestore.rules:58]: NOTE: This needs to be explicitly set for groupQuery to work as per the docs
https://firebase.google.com/docs/firestore/security/rules-query#secure_and_query_documents_based_on_collection_groups

File: firestore.rules [hunk 3]
```diff
@@ -8,6 +8,10 @@ function role(name) {
   return get(/databases/$(database)/documents/roles/$(name)).data
 }
 
+function venueData() {
+  return get(/databases/$(database)/documents/venues/$(venue)).data
+}
```
Comment (mike-lvov) [inline firestore.rules:13]: I needed to move it to use in a new rule

File: src/types/venues.ts (hunk 1 above)
Comment (mike-lvov) [reply src/types/venues.ts:322]: we are going to prepopulate our data with it. So in this environment, yes

File: firestore.rules [hunk 4]
```diff
@@ -48,6 +48,9 @@ service cloud.firestore {
       allow read: if request.auth.uid != null;
       allow write: if true;
     }
+    match /{path=**}/events/{event} {
+      allow read: if true;
+    }
```
Comment (mike-lvov) [inline firestore.rules:53]: NOTE: This needs to be explicitly set for groupQuery to work as per the docs
https://firebase.google.com/docs/firestore/security/rules-query#secure_and_query_documents_based_on_collection_groups

Comment (0xdevalias) [reply firestore.rules:53]: This should probably be `if request.auth.uid != null;` at a minimum. Currently any unauthenticated user on the internet could query this.

File: src/api/events.ts [hunk 5]
```diff
@@ -2,32 +2,24 @@ import firebase from "firebase/app";
...
-): Promise<WithVenueId<WithId<VenueEvent>>[]> =>
-  Promise.all(asArray(venueIdOrIds).map(fetchVenueEvents)).then((result) =>
-    result.flat()
-  );
+  return eventsSnapshot.docs
+    .map((venueEvent) => venueEvent.data())
+    .filter((venueEvent) => venueEvent.venueId) as WithVenueId<
+    WithId<VenueEvent>
+  >[];
+};
```
Comment (0xdevalias) [inline src/api/events.ts:22]: Ignoring the `fetchAllVenueEvents` part which is fine to remove. Why was the `.then` removed + separated like this? It should have just been left inline as it was by the looks of things.

File: src/hooks/events.tsx [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+  venueId,
+}) => {
+  const {
+    sovereignVenueId,
+    isSovereignVenueIdLoading,
+
+    errorMsg: sovereignVenueIdError,
+  } = useSovereignVenueId({
+    venueId,
+  });
```
Comment (0xdevalias) [inline src/hooks/events.tsx:70]: `EventsProvider` is setup to need `venueId` passed to it. Yet where you instantiate it, you're not passing `venueId` to it at all?

File: src/pages/VenuePage/TemplateWrapper.tsx [hunk 7]
```diff
@@ -146,17 +147,19 @@ const TemplateWrapper: React.FC<TemplateWrapperProps> = ({ venue }) => {
...
-          fullscreen={fullscreen}
-          hasBackButton={hasBackButton}
-        >
-          <AnnouncementMessage message={venue.bannerMessage} />
-          {template}
-          <ChatSidebar venue={venue} />
-          <UserProfileModal venue={venue} />
-        </WithNavigationBar>
-      </ReactionsProvider>
+      <EventsProvider>
```
Comment (0xdevalias) [inline src/pages/VenuePage/TemplateWrapper.tsx:150]: Where's the `venueId` prop on `EventsProvider` here?

File: src/types/venues.ts (hunk 1 above)
Comment (0xdevalias) [reply src/types/venues.ts:322]: Makes sense.

**Comment by 0xdevalias** [comment]
@mike-lvov We should think about bringing this change + related out of `env/ohbm` into the main `staging` code branch + cleaning up some related manual aspects of it when we do.

File: src/pages/VenuePage/TemplateWrapper.tsx (hunk 7 above)
Comment (0xdevalias) [reply src/pages/VenuePage/TemplateWrapper.tsx:150]: Seems this was fixed in https://github.com/sparkletown/sparkle/pull/1631

File: src/hooks/events.tsx (hunk 6 above)
Comment (0xdevalias) [reply src/hooks/events.tsx:70]: Seems this was fixed in https://github.com/sparkletown/sparkle/pull/1631

**Comment by 0xdevalias** [comment]
> NOTE: We should also create an index

@mike-lvov An index on what/for what purpose? Doesn't this only use single fields? Firestore indexes all single fields by default.

**Pull request merged**